Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/srvpri/SrvPriManager.java

```diff
@@ -212,47 +219,43 @@ public void write(String str) throws IOException {
         fileReporter.report("Checking cred file...");
 
-        int maxRetry = 4;
-        return Mono.fromCallable(() -> {
+        return retryOnError(RETRY_COUNT, Mono.fromCallable(() -> {
             File authFile = new File(filePath.toString());
-            try {
-                fileReporter.report("Checking: Azure.authenticate(authFile)...");
-                Azure.Authenticated azureAuthenticated = Azure.authenticate(authFile);
-                fileReporter.report("Checking: azureAuthenticated.subscriptions().list()...");
-                azureAuthenticated.subscriptions().list();
-                fileReporter.report("Checking: azureAuthenticated.withDefaultSubscription()...");
-                Azure azure = azureAuthenticated.withDefaultSubscription();
-                fileReporter.report("Checking: resourceGroups().list()...");
-                azure.resourceGroups().list();
-                fileReporter.report("Done.");
-                return true;
-            } catch (Throwable e) {
-                LOGGER.info("=== checkArtifact@SrvPriManager exception: " + e.getMessage());
-                //e.printStackTrace();
-                if (needToRetry(e)) {
-                    // make the retry work
-                    throw e;
-                }
-                return false;
-            }
-        }).retryWhen(Retry.backoff(maxRetry, Duration.ofSeconds(10))).onErrorResume(e -> {
-            fileReporter.report(String.format("Failed to check cred file after %s retries, error: %s", maxRetry, e.getMessage()));
-            return Mono.just(false);
-        });
+            fileReporter.report("Checking: Azure.authenticate(authFile)...");
+            Azure.Authenticated azureAuthenticated = Azure.authenticate(authFile);
+            fileReporter.report("Checking: azureAuthenticated.subscriptions().list()...");
+            azureAuthenticated.subscriptions().list();
+            fileReporter.report("Checking: azureAuthenticated.withDefaultSubscription()...");
+            Azure azure = azureAuthenticated.withDefaultSubscription();
+            fileReporter.report("Checking: resourceGroups().list()...");
+            azure.resourceGroups().list();
+            fileReporter.report("Done.");
+            return true;
+        }), SrvPriManager::needToRetry);
 
+    }
 
+    // TODO: move this to common-lib
+    private static <T> Mono<T> retryOnError(int retryCount, Mono<T> observable, Function<Throwable, Boolean> shouldRetryOnError) {
+        return observable
+                .retryWhen(Retry.withThrowable(flux -> flux.zipWith(Flux.range(1, retryCount + 1), (throwable, count) -> {
+                    if (count < retryCount && (shouldRetryOnError.apply(throwable))) {
+                        return count;
+                    } else {
+                        throw Exceptions.propagate(throwable);
+                    }
+                }).flatMap(i -> Mono.delay(Duration.ofSeconds(((long) i * 10))))));
```

Comment:
I think its good to wait for longer time when retry for a second time, but do we also need to reduce the retry count for create sp? Which used to be 40s but 150s now

Reply:
It is reasonable since 40s is not a precise time where normal role assignment will finish, we will not stick to this magic number